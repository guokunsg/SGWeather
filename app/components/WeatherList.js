// @flow

/** The list display of the weather forecast */

import React, { Component } from 'react'
import { View, Text, Image, FlatList, StyleSheet } from 'react-native'

import getWeatherImageSource from './WeatherImageSource'
import type { LocationForecast } from '../types/WeatherForecast'

const styles = StyleSheet.create({
    list: {
        flex: 1,
    },
    listItem: {
        flex: 1,
        flexDirection: 'column',
        alignItems: 'center',
        padding: 8,
    },
    itemText: {
        flex: 1,
        textAlign: 'center',
        padding: 8,
    },
    itemImage: {
        width: 50,
        height: 50,
    },
    itemTextNoImage: {
        textAlign: 'center',
        textAlignVertical: 'center',
        fontWeight: 'bold',
        height: 50,
    },
})

// Get the Image or Text to display for the weather forecast
function getWeatherImage(forecast: string) {
    const src = getWeatherImageSource(forecast)
    if (src) {
        return <Image style={styles.itemImage} source={src} />
    }
    // Return text if the forecast is not recognized
    return <Text style={styles.itemTextNoImage}>{forecast}</Text>
}

type Props = {
    columns: number,
    data: Array<LocationForecast>
}
export class WeatherList extends Component<Props> {
    renderItem = ({ item }: { item: LocationForecast }) => ( // eslint-disable-line react/no-unused-prop-types
      <View style={styles.listItem}>
        { getWeatherImage(item.forecast) }
        <Text style={styles.itemText}>{item.name}</Text>
      </View>
    );

    keyExtractor = (item: LocationForecast) => item.name

    render() {
        const { data, columns } = this.props
        return (
          <FlatList
            styles={styles.list}
            data={data}
            numColumns={columns}
            keyExtractor={this.keyExtractor}
            renderItem={this.renderItem}
          />
        )
    }
}

export default WeatherList
