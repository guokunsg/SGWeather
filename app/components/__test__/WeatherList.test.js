import React from 'react'
import 'react-native'

import render from 'react-test-renderer'

import WeatherList from '../WeatherList'

// require(...) in WeatherImageSource will cause error.
// Second parameter is the factory function which returns the mocked function that just returns the input
jest.mock('../WeatherImageSource', () => weather => ({ uri: weather }))

describe('Test weather forecast list', () => {
    function createData(name, latitude, longitude, forecast) {
        return { name, latitude, longitude, forecast }
    }

    it('renders empty', () => {
        const tree = render.create(<WeatherList />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('render data', () => {
        const data = []
        data.push(createData('data1', 1.0, 1.0, 'Cloudy'))
        data.push(createData('data2', 1.1, 1.1, 'Partly Cloudy (Day)'))
        data.push(createData('data3', 1.2, 1.2, 'Partly Cloudy (Night)'))
        data.push(createData('data4', 1.3, 1.3, 'Moderate Rain'))
        data.push(createData('data5', 1.4, 1.4, 'Light Rain'))
        data.push(createData('data6', 1.5, 1.5, 'Light Showers'))
        data.push(createData('data7', 1.5, 1.5, 'Showers'))
        data.push(createData('data8', 1.5, 1.5, 'Thundery Showers'))
        data.push(createData('data9', 1.5, 1.5, 'unknown text'))

        const tree = render.create(<WeatherList data={data} />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})
