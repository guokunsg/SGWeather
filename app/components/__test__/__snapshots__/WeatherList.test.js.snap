// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test weather forecast list render data 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "forecast": "Cloudy",
        "latitude": 1,
        "longitude": 1,
        "name": "data1",
      },
      Object {
        "forecast": "Partly Cloudy (Day)",
        "latitude": 1.1,
        "longitude": 1.1,
        "name": "data2",
      },
      Object {
        "forecast": "Partly Cloudy (Night)",
        "latitude": 1.2,
        "longitude": 1.2,
        "name": "data3",
      },
      Object {
        "forecast": "Moderate Rain",
        "latitude": 1.3,
        "longitude": 1.3,
        "name": "data4",
      },
      Object {
        "forecast": "Light Rain",
        "latitude": 1.4,
        "longitude": 1.4,
        "name": "data5",
      },
      Object {
        "forecast": "Light Showers",
        "latitude": 1.5,
        "longitude": 1.5,
        "name": "data6",
      },
      Object {
        "forecast": "Showers",
        "latitude": 1.5,
        "longitude": 1.5,
        "name": "data7",
      },
      Object {
        "forecast": "Thundery Showers",
        "latitude": 1.5,
        "longitude": 1.5,
        "name": "data8",
      },
      Object {
        "forecast": "unknown text",
        "latitude": 1.5,
        "longitude": 1.5,
        "name": "data9",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  styles={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Cloudy",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data1
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Partly Cloudy (Day)",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data2
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Partly Cloudy (Night)",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data3
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Moderate Rain",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data4
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Light Rain",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data5
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Light Showers",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data6
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Showers",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data7
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "Thundery Showers",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data8
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "padding": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "unknown text",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          data9
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Test weather forecast list renders empty 1`] = `
<RCTScrollView
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  styles={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;
